packages/affiliates/libs/uniswap/models.ts

```diff
@@ -0,0 +1,253 @@
+import assert from "assert";
+import { exists, getPositionKey } from "./utils";
+import { ethers } from "ethers";
+
+type CacheType = {
+  [key: string]: string | number | boolean;
+};
+export const Cache = (table = new Map<string | number, CacheType>()) => {
+  function list() {
+    return [...table.values()];
+  }
+  return {
+    has: table.has.bind(table),
+    set: table.set.bind(table),
+    get: table.get.bind(table),
+    delete: table.delete.bind(table),
+    list
+  };
+};
+export type NftPosition = {
+  id?: string;
+  tokenId: string;
+  amount?: string;
+  amount0?: string;
+  amount1?: string;
+  token0?: string;
+  token1?: string;
+  liquidity?: string;
+  feeGrowthInside0LastX128?: string;
+  feeGrowthInside1LastX128?: string;
+  tokensOwed0?: string;
+  tokensOwed1?: string;
+  pool?: string;
+};
+
+// Nft positions are gathered from the nft contract, and require slightly different indexing than the positions in pool contract
+export const NftPositions = (config: any, table: ReturnType<typeof Cache>) => {
+  function makeId(data: NftPosition) {
+    return data.tokenId;
+  }
+  async function create(data: NftPosition) {
+    const id = makeId(data);
+    assert(!(await has(id)), "Position exists");
+    return set({ id, ...data });
+  }
+  async function set(data: NftPosition) {
+    assert(data.id, "Position requires id");
+    await table.set(data.id, { ...data });
+    return data;
+  }
+  async function get(id: string) {
+    assert(await table.has(id), "Position does not exist");
+    return table.get(id);
+  }
+  async function has(id: string) {
+    return table.has(id);
+  }
+  async function forEach(cb: (value: NftPosition, index: number, array: NftPosition[]) => void) {
+    const list: NftPosition[] = (await table.list()) as NftPosition[];
+    list.forEach(cb);
+  }
+  async function update(id: string, data: any = {}) {
+    const position = await get(id);
+    return set({ ...position, ...data });
+  }
+
+  return {
+    create,
+    set,
+    get,
+    has,
+    update,
+    forEach,
+    list: table.list
+  };
+};
+export type Position = {
+  id?: string;
+  operator: string;
```

Comment:
Same question above for `tableId vs id` but this time with `operator and id`. Maybe there's something about `typescript` i'm missing.

Reply:
position is slightly more complex, its a hash of operator, lower tick and upper tick. in this case you have to derive the key, which you can see in the utils file:
```js
export function getPositionKey(address: string, lowerTick: string, upperTick: string): string {
  return ethers.utils.keccak256(
    ethers.utils.solidityPack(["address", "int24", "int24"], [address, lowerTick, upperTick])
  );
}
```

I think in this case it makes sense to have an id